lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs

```diff
@@ -265,3 +324,140 @@ errorHandlingSpec = describe "liftHandler and toServerError" $ do
 
         runHandler (getNetworkClock (error "bomb") True)
             `shouldReturn` Left expectedErr
+
+listStakeKeysSpec :: Spec
+listStakeKeysSpec = do
+    describe "Listing stake keys" $ do
+        it "no rewards => sum of all keys' stake == utxo balance" $ do
+            property prop_listStakeKeysBalance
+        it "ours and foreign keys are disjoint" $ do
+            property prop_listStakeKeysDisjoint
+
+prop_listStakeKeysDisjoint
+    :: PrettyShow UTxO
+    -> SaneOurStakeKeysInput
+    -> (Map RewardAccount Coin)
+    -> Property
+prop_listStakeKeysDisjoint (PrettyShow utxo) (SaneOurStakeKeysInput ours) m =
+    let
+        res@ApiStakeKeys{_ours, _foreign, _none } = runIdentity $
+            listStakeKeys' utxo accountOfAddress (const $ pure m) ours
+
+    in
+        checkCoverage $ cover 35 (isNonTrivial res) "non-trivial" $
+            keys _ours `disjoint` keys _foreign
+  where
+    labeledCheck notF f a b = counterexample
+        (show a <> " " <> notF <> " " <> show b)
+        (property $ f a b)
+
+    disjoint = labeledCheck "is not disjoint from" Set.disjoint
+
+prop_listStakeKeysBalance
+    :: PrettyShow UTxO
+    -> SaneOurStakeKeysInput
+    -> Property
+prop_listStakeKeysBalance (PrettyShow utxo) (SaneOurStakeKeysInput ours) =
+    let
+        res@ApiStakeKeys{_ours, _foreign, _none } = runIdentity $
+            listStakeKeys' utxo accountOfAddress (const $ pure mempty) ours
+
+    in
+        checkCoverage $ cover 35 (isNonTrivial res) "non-trivial" $
+            mconcat
+                [ totalStake _ours
+                , totalStake _foreign
+                , totalStake [_none]
+                ]
+                ===
+            TokenBundle.getCoin (balance utxo)
+
+-- | Type for where the reward accounts and indices are unique.
+--
+-- This is required for some properties, and does not seem worthwhile to protect
+-- against.
+newtype SaneOurStakeKeysInput
```

Comment:
If unique accounts and indices are a precondition of `listStakeKeys'`, then perhaps this wrapper type should be used for the `ourKeys` parameter of that function?

But on the other hand, the `ourKeys` list is always of length 1, so unique by definition...